File: OPEN_THIS/SETUP/setup_loader.py
#!/usr/bin/env python3
"""
Setup Configuration Loader (Python)
Loads and validates configuration from setup_variable.json
"""

import json
import os
import sys
from pathlib import Path
from typing import Dict, List, Optional, Any


class SetupConfigLoader:
    """Loader for setup_variable.json configuration"""
    
    def __init__(self, config_path: Optional[str] = None):
        """
        Initialize configuration loader
        
        Args:
            config_path: Optional path to setup_variable.json
        """
        if config_path is None:
            # Default to OPEN_THIS/SETUP directory
            script_dir = Path(__file__).parent
            config_path = script_dir / 'setup_variable.json'
        
        self.config_path = Path(config_path)
        self.config: Optional[Dict[str, Any]] = None
        self.loaded = False
    
    def load(self) -> Dict[str, Any]:
        """
        Load configuration from file
        
        Returns:
            Configuration dictionary
        """
        if self.loaded and self.config:
            return self.config
        
        try:
            if not self.config_path.exists():
                raise FileNotFoundError(f'Configuration file not found: {self.config_path}')
            
            with open(self.config_path, 'r', encoding='utf-8') as f:
                self.config = json.load(f)
            
            # Sync simplified top-level fields to internal config structure
            if 'YOUR_IP_ADDRESS' in self.config:
                self.config['server']['host'] = self.config['YOUR_IP_ADDRESS']
                ip_parts = self.config['YOUR_IP_ADDRESS'].rsplit('.', 1)
                self.config['network']['subnet'] = ip_parts[0] + '.x'
                port = self.config.get('SERVER_PORT', 3001)
                self.config['network']['allowed_origins'][0] = f"http://{self.config['YOUR_IP_ADDRESS']}:{port}"
            if 'SERVER_PORT' in self.config:
                self.config['server']['port'] = self.config['SERVER_PORT']
            if 'MATLAB_PATH' in self.config:
                self.config['matlab']['installation_paths'] = [self.config['MATLAB_PATH']]
            if 'PYTHON_PATH' in self.config:
                self.config['python']['executable'] = self.config['PYTHON_PATH']
            
            self.loaded = True
            print(f'✅ Loaded configuration from {self.config_path.name}')
            return self.config
        
        except Exception as e:
            print(f'❌ Failed to load configuration: {e}', file=sys.stderr)
            raise
    
    def get_matlab_paths(self) -> List[str]:
        """Get MATLAB installation paths"""
        config = self.load()
        return config.get('matlab', {}).get('installation_paths', [])
    
    def get_python_executable(self) -> str:
        """Get Python executable path"""
        config = self.load()
        return config.get('python', {}).get('executable', 'python')
    
    def get_server_config(self) -> Dict[str, Any]:
        """Get server configuration"""
        config = self.load()
        server = config.get('server', {})
        host = server.get('host')
        port = server.get('port', 3001)
        
        if not host:
            raise ValueError('Server host not configured in setup_variable.json. Please set server.host to your PC\'s IP address.')
        
        return {
            'host': host,
            'port': port,
            'url': f'http://{host}:{port}',
            'websocket': {
                'enabled': server.get('websocket', {}).get('enabled', True),
                'path': server.get('websocket', {}).get('path', '/ws'),
                'url': f'ws://{host}:{port}{server.get("websocket", {}).get("path", "/ws")}'
            }
        }
    
    def get_hfss_config(self) -> Dict[str, Any]:
        """Get HFSS configuration"""
        config = self.load()
        return config.get('hfss', {})
    
    def get_project_paths(self) -> Dict[str, Path]:
        """Get project paths configuration"""
        config = self.load()
        paths = config.get('paths', {})
        
        # Get project root (usually current directory)
        project_root = Path.cwd()
        
        return {
            'project_root': project_root / paths.get('project_root', '.'),
            'uploads_dir': project_root / paths.get('uploads_dir', './uploads'),
            'gnd_files_dir': project_root / paths.get('gnd_files_dir', './uploads/gnd_files'),
            'config_dir': project_root / paths.get('config_dir', './config'),
            'scripts_dir': project_root / paths.get('scripts_dir', './scripts'),
            'test_files_dir': project_root / paths.get('test_files_dir', './test_files')
        }
    
    def get_performance_settings(self) -> Dict[str, Any]:
        """Get performance settings"""
        config = self.load()
        return config.get('performance', {
            'cache_ttl_ms': 1000,
            'websocket_heartbeat_ms': 2000,
            'status_polling_interval_ms': 3000,
            'max_file_upload_mb': 50
        })
    
    def validate(self) -> Dict[str, Any]:
        """
        Validate configuration
        
        Returns:
            Dictionary with validation result
        """
        errors = []
        warnings = []
        
        try:
            config = self.load()
            
            # Check required fields
            if 'matlab' not in config or 'installation_paths' not in config.get('matlab', {}):
                errors.append('Missing MATLAB installation paths')
            
            if 'server' not in config or not config.get('server', {}).get('host') or not config.get('server', {}).get('port'):
                errors.append('Missing server configuration')
            
            if 'python' not in config or not config.get('python', {}).get('executable'):
                errors.append('Missing Python executable configuration')
            
            # Check if MATLAB exists
            matlab_paths = self.get_matlab_paths()
            valid_matlab_path = None
            for matlab_path in matlab_paths:
                if Path(matlab_path).exists():
                    valid_matlab_path = matlab_path
                    break
            
            if not valid_matlab_path:
                warnings.append('MATLAB installation not found in configured paths')
            
            return {
                'valid': len(errors) == 0,
                'errors': errors,
                'warnings': warnings
            }
        
        except Exception as e:
            return {
                'valid': False,
                'errors': [str(e)],
                'warnings': []
            }
    
    def get_config(self) -> Dict[str, Any]:
        """Get full configuration object"""
        return self.load()


# Singleton instance
_setup_config_instance = None


def get_setup_config(config_path: Optional[str] = None) -> SetupConfigLoader:
    """
    Get singleton instance of SetupConfigLoader
    
    Args:
        config_path: Optional path to configuration file
    
    Returns:
        SetupConfigLoader instance
    """
    global _setup_config_instance
    
    if _setup_config_instance is None:
        _setup_config_instance = SetupConfigLoader(config_path)
    
    return _setup_config_instance


# Convenience functions
def get_matlab_paths() -> List[str]:
    """Get MATLAB installation paths"""
    return get_setup_config().get_matlab_paths()


def get_server_url() -> str:
    """Get server URL"""
    return get_setup_config().get_server_config()['url']


def get_python_executable() -> str:
    """Get Python executable path"""
    return get_setup_config().get_python_executable()


if __name__ == '__main__':
    # Test configuration loading
    loader = get_setup_config()
    
    print('\n=== Setup Configuration ===')
    print(f'Config file: {loader.config_path}')
    
    # Validate
    validation = loader.validate()
    print(f'\nValidation: {"✅ PASSED" if validation["valid"] else "❌ FAILED"}')
    
    if validation['errors']:
        print('\nErrors:')
        for error in validation['errors']:
            print(f'  ❌ {error}')
    
    if validation['warnings']:
        print('\nWarnings:')
        for warning in validation['warnings']:
            print(f'  ⚠️  {warning}')
    
    # Display configuration
    print('\n=== Configuration Summary ===')
    print(f'Server: {get_server_url()}')
    print(f'Python: {get_python_executable()}')
    print(f'MATLAB paths: {len(get_matlab_paths())} configured')
    
    server_config = loader.get_server_config()
    print(f'WebSocket: {server_config["websocket"]["url"]}')

File: OPEN_THIS/SETUP/setup_loader.js
/**
 * Setup Configuration Loader (Node.js)
 * Loads and validates configuration from setup_variable.json
 */

const fs = require('fs');
const path = require('path');

class SetupConfigLoader {
    constructor() {
        this.configPath = path.join(__dirname, 'setup_variable.json');
        this.config = null;
        this.loaded = false;
    }

    /**
     * Load configuration from file
     * @returns {Object} Configuration object
     */
    load() {
        if (this.loaded && this.config) {
            return this.config;
        }

        try {
            if (!fs.existsSync(this.configPath)) {
                throw new Error(`Configuration file not found: ${this.configPath}`);
            }

            const configData = fs.readFileSync(this.configPath, 'utf-8');
            this.config = JSON.parse(configData);
            
            // Sync simplified top-level fields to internal config structure
            if (this.config.YOUR_IP_ADDRESS) {
                this.config.server.host = this.config.YOUR_IP_ADDRESS;
                this.config.network.subnet = this.config.YOUR_IP_ADDRESS.substring(0, this.config.YOUR_IP_ADDRESS.lastIndexOf('.')) + '.x';
                this.config.network.allowed_origins[0] = `http://${this.config.YOUR_IP_ADDRESS}:${this.config.SERVER_PORT || 3001}`;
            }
            if (this.config.SERVER_PORT) {
                this.config.server.port = this.config.SERVER_PORT;
            }
            if (this.config.MATLAB_PATH) {
                this.config.matlab.installation_paths = [this.config.MATLAB_PATH];
            }
            if (this.config.PYTHON_PATH) {
                this.config.python.executable = this.config.PYTHON_PATH;
            }
            
            this.loaded = true;

            console.log('✅ Loaded configuration from setup_variable.json');
            return this.config;
        } catch (error) {
            console.error('❌ Failed to load configuration:', error.message);
            throw error;
        }
    }

    /**
     * Get MATLAB installation paths
     * @returns {Array} Array of MATLAB installation paths
     */
    getMatlabPaths() {
        const config = this.load();
        return config.matlab.installation_paths || [];
    }

    /**
     * Get server configuration
     * @returns {Object} Server configuration
     */
    getServerConfig() {
        const config = this.load();
        const host = config.server?.host;
        const port = config.server?.port || 3001;
        
        if (!host) {
            throw new Error('Server host not configured in setup_variable.json. Please set server.host to your PC\'s IP address.');
        }
        
        return {
            host: host,
            port: port,
            url: `http://${host}:${port}`,
            websocket: {
                enabled: config.server?.websocket?.enabled !== false,
                path: config.server?.websocket?.path || '/ws',
                url: `ws://${host}:${port}${config.server?.websocket?.path || '/ws'}`
            }
        };
    }

    /**
     * Get Python executable path
     * @returns {string} Python executable path
     */
    getPythonExecutable() {
        const config = this.load();
        return config.python.executable || 'python';
    }

    /**
     * Get HFSS process names
     * @returns {Array} Array of HFSS process names
     */
    getHfssProcessNames() {
        const config = this.load();
        return config.hfss.process_names || [
            'ansysedt.exe',
            'anshfss.exe',
            'ansysli_server.exe'
        ];
    }

    /**
     * Get network configuration
     * @returns {Object} Network configuration
     */
    getNetworkConfig() {
        const config = this.load();
        return {
            allowedOrigins: config.network.allowed_origins || [],
            corsEnabled: config.network.cors_enabled !== false
        };
    }

    /**
     * Get performance settings
     * @returns {Object} Performance settings
     */
    getPerformanceSettings() {
        const config = this.load();
        return config.performance || {
            cache_ttl_ms: 1000,
            websocket_heartbeat_ms: 2000,
            status_polling_interval_ms: 3000,
            max_file_upload_mb: 50
        };
    }

    /**
     * Get project paths
     * @returns {Object} Project paths configuration
     */
    getProjectPaths() {
        const config = this.load();
        const projectRoot = process.cwd();
        
        return {
            projectRoot: path.resolve(config.paths.project_root || projectRoot),
            uploadsDir: path.resolve(config.paths.uploads_dir || path.join(projectRoot, 'uploads')),
            gndFilesDir: path.resolve(config.paths.gnd_files_dir || path.join(projectRoot, 'uploads', 'gnd_files')),
            configDir: path.resolve(config.paths.config_dir || path.join(projectRoot, 'config')),
            scriptsDir: path.resolve(config.paths.scripts_dir || path.join(projectRoot, 'scripts')),
            testFilesDir: path.resolve(config.paths.test_files_dir || path.join(projectRoot, 'test_files'))
        };
    }

    /**
     * Validate configuration
     * @returns {Object} Validation result {isValid: boolean, errors: Array, warnings: Array}
     */
    validate() {
        const errors = [];
        const warnings = [];

        try {
            const config = this.load();

            // Check required top-level fields
            if (!config.YOUR_IP_ADDRESS) {
                errors.push('❌ YOUR_IP_ADDRESS is missing');
            }
            if (!config.SERVER_PORT) {
                errors.push('❌ SERVER_PORT is missing');
            }
            if (!config.MATLAB_PATH) {
                errors.push('❌ MATLAB_PATH is missing');
            }
            if (!config.PYTHON_PATH) {
                errors.push('❌ PYTHON_PATH is missing');
            }

            // Check MATLAB path exists
            if (config.MATLAB_PATH && !fs.existsSync(config.MATLAB_PATH)) {
                errors.push(`❌ MATLAB not found at: ${config.MATLAB_PATH}`);
            }

            // Check Python path exists
            if (config.PYTHON_PATH && !fs.existsSync(config.PYTHON_PATH)) {
                errors.push(`❌ Python not found at: ${config.PYTHON_PATH}`);
            }

            // Check IP format
            if (config.YOUR_IP_ADDRESS) {
                const ipRegex = /^(\d{1,3}\.){3}\d{1,3}$/;
                if (!ipRegex.test(config.YOUR_IP_ADDRESS)) {
                    warnings.push(`⚠️  IP address format may be invalid: ${config.YOUR_IP_ADDRESS}`);
                }
            }

            return {
                isValid: errors.length === 0,
                errors: errors,
                warnings: warnings
            };
        } catch (error) {
            return {
                isValid: false,
                errors: [`❌ Failed to load configuration: ${error.message}`],
                warnings: []
            };
        }
    }

    /**
     * Get full configuration object
     * @returns {Object} Complete configuration
     */
    getConfig() {
        return this.load();
    }
}

// Singleton instance
const setupConfig = new SetupConfigLoader();

module.exports = setupConfig;
